### DISCUSSION TITLE
Consider run service as non-root user.

### DISCUSSION THREAD

**Discussion by libook** [AbdBarho/stable-diffusion-webui-docker#238] [Ideas]
Thank AbdBarho for maintaining this.
I just tried this, and it works very well. Pictures were generated with root user and root group. I have to do `chown` to operating them as normal users. So I have added lines into `services/AUTOMATIC1111/Dockerfile` before `WORKDIR` to let it run as a non-root user:

```Dockerfile
RUN \
  # There is already a group named users with GID 100
  useradd -m -u 1000 -g users -o -s /bin/sh user && \
  mkdir /output /data && \
  chown -R user:users /output /data $ROOT
USER user
```

(Also use `ARG` to customize UID and GID.)

And I have modified `services/AUTOMATIC1111/entrypoint.sh` to switch home directory:

```diff
-MOUNTS["/root/.cache"]="/data/.cache"
+MOUNTS["/home/user/.cache"]="/data/.cache"
```

Would this be a good idea?

**Comment by AbdBarho** [comment]
@libook I agree it is a good idea to run as non-root. 

However, this will cause problems when working with extensions, installing additional dependencies, or wanting to customize the behaviour.

**Reply by libook** [reply]
OK I got it. Well, anyone can do a hack with this information if they want to run as non-root.

**Reply by putnam** [reply]
I run the whole stack as non-root without Docker, and everything works fine, so why does running it as non-root in this image not work? It seems really dangerous considering the situation (lots of people running untrusted code via pickled ckpt files, dozens of untrusted repos being downloaded and executed on the fly, the auto1111 repo itself being fairly untrustable, etc). You can use a technique like LSIO does with S6 overlay. The image should be built with some base unpriv user like "abc" owning everything. Then at startup you just change the UID/GID to whatever is provided via environment variables PUID/PGID

Example:
https://github.com/linuxserver/docker-baseimage-alpine/blob/master/root/etc/s6-overlay/s6-rc.d/init-adduser/run

This is in very wide use, works fine

**Reply by AbdBarho** [reply]
@putnam can you install apt packages with a non-root user without giving them root privileges? it would be great if you could tell me how so I can integrate it.

**Reply by putnam** [reply]
Hm, I haven't seen my auto1111 install try to install packages with apt. How does that work? I use extensions and everything

**Reply by AbdBarho** [reply]
for example, in case you need to install `ffmpeg` for some video stuff

**Reply by putnam** [reply]
That's something you would do in the build step; docker images aren't meant to be long-lived like that. You are storing the data separately so the whole image is easily rebuilt without affecting it.

**Reply by DevilaN** [reply]
It would be the case when you have application that is docker ready. A1111 is certainly not prepared this way. Extensions and scripts require installing external dependencies "by hand" to work. We could not cover / support each and every possible extension right now. For a time being it has to be like that right now.
Judging by the issue history for this repository, most of users have almost no idea about docker, so extending Dockerfile and building special images each time you need to try some new extension would be great source of frustration for users.
Yeah, it would be ideal situation when it is running that way, but world is - sadly - not ideal place.

**Reply by putnam** [reply]
I don't think I follow. If today you have to shell in and install packages by hand to enable certain extensions, this is still possible with an image that runs the actual init/script as a non-root user. You can still shell in as root and manually install things. But again this is something you would accomplish by modifying the build and rebuilding -- after all your readme is encouraging users to rebuild the image completely every time it's run anyway.

This is not any more or less convenient than it is today, and the way it is today is unnecessarily risky, right?

What are some example extensions that require these efforts? Maybe I can help come up with a solution.

The idea of "hand installing" apt packages by the user doesn't even make sense with docker. As soon as you restart the container those packages are gone. You have to do this in the image building step (or perhaps in your init script before you drop privs). So if users need to be able to "hand install" or hand-enable apt packages you should just provide a way to do that in the build stage so it's actually more permanent.

**Reply by putnam** [reply]
Reading the replies again I feel there is a miscommunication. I am not saying to set this up to be run with --user. I am saying to use S6 overlay or a similar technique to run Auto1111/Invoke as a non-root user. Important distinction.

Ideally, things would work with --user, but as you suggested apt packages would be effectively frozen.

By using S6 overlay (or similar) you can still have the image running as root but run the Auto service in a supervised way as an unprivileged user. You don't even need the whole S6 stack. You just need to setup an entrypoint that looks for some environment variables for the uid/gid to run the auto1111 process, and make sure to chown the data folders before running to avoid any permissions issues. Same thing S6 does, linked above. User can still shell in as root and install packages and manipulate the live filesystem.

This technique is literally designed to work around apps that aren't as "docker ready", so it's ideal for this situation.

**Reply by AbdBarho** [reply]
@putnam thanks for the suggestion, I am not familiar with the solution you are proposing here, maybe you can help me a bit, there are 3 crucial points for this repo:

* is this only a change to the docker / compose setup? (in contrast to the user having to install additional software to make this work on their machine)
* is the installation of apt packages possible without modifying the dockerfile?
* does this work *seamlessly* on windows?

If the answer for any of these question is 'no', I am afraid we cannot move forward with this. As @DevilaN mentioned, many of our users are not familiar with docker, and my aim with this repo is to make the usage as easy as possible.

> after all your readme is encouraging users to rebuild the image completely every time it's run anyway.

The reason for this is because I am certain that some users (who are unfamiliar with docker) will try to update the container with the help of some blog post they found online, and then try to run without using `--build` flag and be very confused why their changes are not applied, which in turn will lead to many more unnecessary issues & headaches. 


This comes from my own experience when I was first learning how to use docker and docker compose.


Of course, any confident docker user will know that rebuilding every time is not necessary.

**Reply by putnam** [reply]
I still don't see any examples where you need to install apt packages as the user, ever. Can you provide examples?

I saw [this issue](https://github.com/AbdBarho/stable-diffusion-webui-docker/issues/308) that seems possibly related to what you're referring to. They install ffmpeg [using pip](https://pypi.org/project/imageio-ffmpeg/), which doesn't require root and does not use apt. As long as the user has write access to a folder they can install pip packages to the virtual environment. It's just writing to the venv folder.

There is no method for automatic extensions to install anything with apt. They're using pip.

Also, the requirement conflict you see there is a problem entirely of the extension author's own creation. They have since fixed it. At the moment there is no concept of a separate venv for extensions so them pinning a module version shared with the main app is obviously not a good idea.

**Reply by putnam** [reply]
Regarding keeping it simple for new users and making sure everything works on Windows: definitely can be done. I'm down to help make it work on all platforms.

The question is whether it's any different to pursue the USER directive (--user) style of execution, which is the more modern and preferred way, or to use a privilege dropping strategy like S6 overlay offers. If it's possible to do the former, that would be even better!

I just want to know if there are any edge cases I'm not thinking of that *actually require* root access.

**Reply by DevilaN** [reply]
A1111 is a mess. Extensions are equally messy. TBH it is miracle that it holds till now and works.
From what I've observed, if something is not working it is up to you to install requirements, packets, external software.

For example I am using extension that creates SVG from creations (vector art models work with this extension fine). This extension relies on external binary: `potrace`. I know almost nothing about python, but readme for extension stated you need to have potrace binary for it to work.

@putnam: your example script only creates user and it's home directory. Yeah, running as root in container is not the best idea, but nonetheless worst either. I am afraid that covering each "exception" might take much more time than expected.

**Reply by putnam** [reply]
What example script? That's .. not .. that's just one part of the s6 overlay system. Did you read about S6 on its readme in their repo? Do you want to see how they drop privileges? [There's a helper tool](https://github.com/just-containers/s6-overlay#dropping-privileges) that you run in your entrypoint or service definition.

The extension I think you mean, which I have to guess: https://github.com/GeorgLegato/Txt2Vectorgraphics

Users can *still use these things* -- they can just drop potrace into the scripts folder rather than run apt. Or they could run apt if you used privilege dropping since they can get a shell up. Or they could use startup.sh. It's no more difficult than what they already are doing. Giving root to this stuff is fighting fish with a hand grenade.

OP already has a solution that works fine and I've already written my own that works completely fine and securely as well so I'm going to leave it here. The problem is this is currently the more popular image (but isn't published on Dockerhub or anywhere?). Security is critically important with a project like this considering the underlying service is supposedly run by channers globbing random scripts together from who-knows-where. I feel like it's really not much of a challenge to make it work a little more safely. If this is widely installed on a lot of machines it'll be very interesting if someone wants to have some fun. Good luck!

**Reply by libook** [reply]
The maintainer wants to reduce the requirement for users to know about docker. This is a responsible approach to the product and its users.
Then I realized that perhaps the two could be non-contradictory.
The solution I provided is not limited to non-root users only. If you use ARG (or other features) to configure UIDs and GIDs, you can configure them as root users and root groups by default. For basic users, the service will still run as root out of the box. For advanced users, you can configure the parameters in `docker-compose.yml` to run the service as a non-root user. Of course, advanced users can easily do this themselves, they just don't need to maintain files other than `docker-compose.yml` on their own.

These are some of the inspirations I would like to share.

**Reply by DevilaN** [reply]
@putnam: I agree with you. And most of things require additional research to work. I know docker, you know docker and few other people know docker too. @libook summed up what I am thinking. In the end it is up to @AbdBarho :-)

**Reply by AbdBarho** [reply]
we can maybe start with the other services, like `invoke` which does not require external dependencies.

I have unfortunately less time now to work on this since I am currently busy with [open assistant](https://github.com/LAION-AI/Open-Assistant), so it might drag along for a while, in case any one wants to contribute, feel free to do so!

**Reply by izackv** [reply]
We can force the docker to use our own user ID and group ID without changing the internal use of the root user.
In the `docker-compose.yml` file, under the  `x-base_service` service template, we can add the following:
`user: "${UID}:${GID}"`
Then to start the service, we should use the following:
`UID=${UID} GID=${GID} docker compose --profile [ui] up --build`
To simplify it, we can add a bash script to run the docker-compose command, get the service name as a parameter, and run it all using the above docker-compose command.

**Reply by AbdBarho** [reply]
Would also need to find a way for powershell / cmd users.

**Reply by izackv** [reply]
I believe that running a Linux docker (on a Windows machine) with a UID of 1000 should work without issues.
If this is the case, we can set the user on the `x-base-service` to `user: "${UID}"`, and in the bash to set the `${UID}` to the current user ID, while in the power shell, just set it to `1000`. 
Unfortunately, I can't check it myself, since I don't have access to any windows machine, but it sounds like a quick test.
--- Edit:
If anyone can verify that windows will work with the above solution, I can write it, test it on Linux and make a pull request.

**Reply by izackv** [reply]
I don't think that Windows should be a problem.
I just changed the yml file with the following addition:
```
x-base_service: &base_service
    user: "${USER_ID:-0:0}"
```
This way, the user can set the USER_ID environment variable.
But if no one sets this variable, Docker will use the default (0:0), which should be OK.

To run the file from windows, use the current way.
To run the file from Linux and force the current user ID on the docker, just use the following starting bash script:
```
#!/bin/bash
 if [ "$#" -ne  "1" ]
   then
     echo "usage: $ ./run.sh [ download | auto | auto-cpu | invoke | sygil | sygil-sl ]"
 else
     USER_ID=$(id -u):$(id -g) docker compose --profile $1 up --build
 fi
```

Since I need this change for myself anyway, I created a pull request with those changes.

**Reply by AbdBarho** [reply]
we might need a migration script that allows access for "old" files to non-root users.

**Reply by izackv** [reply]
My code set the container permissions to the current user, but we must fix some issues using a non-root user.
I will investigate that.
We might need to change the `entrypoint.sh` script and others.

**Reply by AbdBarho** [reply]
please don't invest a huge amount of time in this issue, it is secondary for this repo. 

Unless you want it for yourself.
